### DISCUSSION TITLE
Docker Windows: BB asks to create Admin user after Cleaning up unused images

### DISCUSSION THREAD

**Discussion by sergey-dme** [Budibase/budibase#2225] [Help] [answered]
Hi All,

There are two more unused bb images appears (app and worker) in the Docker every time after update and they take some disk space. I tried to remove unused images using "Clean Up" in Docker, after that BB asks to create new Admin user like it is just installed (I guess it happens because some flag file or something gets removed). 
So, first time I was confused, like - "omg.. really? all my data lost after cleaning up..."  
But fortunately not, after post-cleaning-up' "admin user creation" it redirects to builder login page where I can log on using my existing credentials. 
Long story short - is there a way to disable duplication (storing) unused BB images after update?

**Comment by shogunpurple** [comment, marked as answer]
Hi @sergey-dme - thanks for raising. Are you updating through the budibase UI, or are you using `docker-compose` commands?

**Reply by sergey-dme** [reply]
hey @shogunpurple I always use `hosting --update` in cli

**Reply by shogunpurple** [reply]
Ah, that's probably why. When you update through the UI, watchtower (the app that manages the updates) will clean up your images. 

If you want to update all your images with the CLI and clean up old containers and images, you can run:
```
hosting --update
docker system prune
```

**Reply by sergey-dme** [reply]
@shogunpurple Thanks you for the advice! I will update through the UI next time! Is restart required after it?

**Reply by shogunpurple** [reply]
Nope, it should just work. One thing to mention is that we do not have very good feedback in the UI for updates, so click the update button, and just give it a bit of time to update, then when you refresh the page you should see the updated version.

**Reply by sergey-dme** [reply]
@shogunpurple That is valuable information. Thanks again.

**Reply by sergey-dme** [reply]
@shogunpurple Hi Martin, updating trough UI does not work. I checked watchover output in docker, there are some authorization errors:

```
time="2021-08-04T21:16:21Z" level=debug msg="Valid token found."
time="2021-08-04T21:16:21Z" level=info msg="Updates triggered by HTTP API request."
time="2021-08-04T21:16:21Z" level=debug msg="Checking containers for updated images"
time="2021-08-04T21:16:21Z" level=debug msg="Retrieving running containers"

time="2021-08-04T21:16:21Z" level=debug msg="Trying to load authentication credentials." container=/bbapps image="sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"
time="2021-08-04T21:16:21Z" level=debug msg="No credentials for sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128 found" config_file=/config.json
time="2021-08-04T21:16:21Z" level=debug msg="Got image name: sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"
time="2021-08-04T21:16:21Z" level=debug msg="Checking if pull is needed" container=/bbapps image="sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"
time="2021-08-04T21:16:21Z" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"

time="2021-08-04T21:16:22Z" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2021-08-04T21:16:22Z" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2021-08-04T21:16:22Z" level=debug msg="Setting scope for auth token" image=sha256 scope="repository:library/sha256:pull"
time="2021-08-04T21:16:22Z" level=debug msg="No credentials found."

time="2021-08-04T21:16:23Z" level=debug msg="Parsing image ref" host=index.docker.io image=sha256 normalized="docker.io/library/sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128" tag=a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128
time="2021-08-04T21:16:23Z" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/library/sha256/manifests/a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"

time="2021-08-04T21:16:23Z" level=warning msg="Could not do a head request for \"sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128\", falling back to regular pull." container=/bbapps image="sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"
time="2021-08-04T21:16:23Z" level=warning msg="Reason: registry responded to head request with \"401 Unauthorized\", auth: \"Bearer realm=\\\"https://auth.docker.io/token\\\",service=\\\"registry.docker.io\\\",scope=\\\"repository:library/sha256:pull\\\",error=\\\"insufficient_scope\\\"\"" container=/bbapps image="sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"
time="2021-08-04T21:16:23Z" level=debug msg="Pulling image" container=/bbapps image="sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128"

time="2021-08-04T21:16:26Z" level=debug msg="Error pulling image sha256:a58a2d2857de889bfbda534f4c75c06bb25d7cc5ac3194ffdf01542dbe4b9128, Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"
time="2021-08-04T21:16:26Z" level=info msg="Unable to update container \"/bbapps\": Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied. Proceeding to next."
time="2021-08-04T21:16:26Z" level=debug msg="Trying to load authentication credentials." container=/bbworker image="sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"
time="2021-08-04T21:16:26Z" level=debug msg="No credentials for sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f found" config_file=/config.json
time="2021-08-04T21:16:26Z" level=debug msg="Got image name: sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"
time="2021-08-04T21:16:26Z" level=debug msg="Checking if pull is needed" container=/bbworker image="sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"
time="2021-08-04T21:16:26Z" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"

time="2021-08-04T21:16:26Z" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2021-08-04T21:16:26Z" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2021-08-04T21:16:26Z" level=debug msg="Setting scope for auth token" image=sha256 scope="repository:library/sha256:pull"
time="2021-08-04T21:16:26Z" level=debug msg="No credentials found."

time="2021-08-04T21:16:26Z" level=debug msg="Parsing image ref" host=index.docker.io image=sha256 normalized="docker.io/library/sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f" tag=e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f
time="2021-08-04T21:16:26Z" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/library/sha256/manifests/e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"

time="2021-08-04T21:16:27Z" level=warning msg="Could not do a head request for \"sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f\", falling back to regular pull." container=/bbworker image="sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"
time="2021-08-04T21:16:27Z" level=warning msg="Reason: registry responded to head request with \"401 Unauthorized\", auth: \"Bearer realm=\\\"https://auth.docker.io/token\\\",service=\\\"registry.docker.io\\\",scope=\\\"repository:library/sha256:pull\\\",error=\\\"insufficient_scope\\\"\"" container=/bbworker image="sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"
time="2021-08-04T21:16:27Z" level=debug msg="Pulling image" container=/bbworker image="sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f"

time="2021-08-04T21:16:29Z" level=debug msg="Error pulling image sha256:e3db5643b9f08dd7c5f4c89c5621e08f06a44169862833fe8c1c0190b4ef240f, Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"
time="2021-08-04T21:16:29Z" level=info msg="Unable to update container \"/bbworker\": Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied. Proceeding to next."
time="2021-08-04T21:16:29Z" level=debug msg="Session done: 2 scanned, 0 updated, 2 failed"`
```

**Reply by shogunpurple** [reply]
@sergey-dme what does your docker-compose.yaml look like?

**Reply by sergey-dme** [reply]
@shogunpurple As follows:

```
version: "3"

# optional ports are specified throughout for more advanced use cases.

services:
  app-service:
    restart: always
    image: budibase/apps
    container_name: bbapps
    ports:
      - "${APP_PORT}:4002"
    environment:
      SELF_HOSTED: 1
      COUCH_DB_URL: http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984
      WORKER_URL: http://worker-service:4003
      MINIO_URL: http://minio-service:9000
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      INTERNAL_API_KEY: ${INTERNAL_API_KEY}
      BUDIBASE_ENVIRONMENT: ${BUDIBASE_ENVIRONMENT}
      PORT: 4002
      JWT_SECRET: ${JWT_SECRET}
      LOG_LEVEL: info
      SENTRY_DSN: https://a34ae347621946bf8acded18e5b7d4b8@o420233.ingest.sentry.io/5338131
      ENABLE_ANALYTICS: "true"
      REDIS_URL: redis-service:6379
      REDIS_PASSWORD: ${REDIS_PASSWORD}
    volumes:
      - ./logs:/logs
    depends_on:
      - worker-service
      - redis-service

  worker-service:
    restart: always
    image: budibase/worker
    container_name: bbworker
    ports:
      - "${WORKER_PORT}:4003"
    environment:
      SELF_HOSTED: 1
      PORT: 4003
      CLUSTER_PORT: ${MAIN_PORT}
      JWT_SECRET: ${JWT_SECRET}
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      MINIO_URL: http://minio-service:9000
      COUCH_DB_USERNAME: ${COUCH_DB_USER}
      COUCH_DB_PASSWORD: ${COUCH_DB_PASSWORD}
      COUCH_DB_URL: http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984
      INTERNAL_API_KEY: ${INTERNAL_API_KEY}
      REDIS_URL: redis-service:6379
      REDIS_PASSWORD: ${REDIS_PASSWORD}
    volumes:
      - ./logs:/logs
    depends_on:
      - redis-service
      - minio-service
      - couch-init

  minio-service:
    restart: always
    image: minio/minio
    volumes:
      - minio_data:/data
    ports:
      - "${MINIO_PORT}:9000"
    environment:
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      MINIO_BROWSER: "off"
    command: server /data
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3

  proxy-service:
    restart: always
    image: envoyproxy/envoy:v1.16-latest
    volumes:
      - ./envoy.yaml:/etc/envoy/envoy.yaml
    ports:
      - "${MAIN_PORT}:10000"
    depends_on:
      - minio-service
      - worker-service
      - app-service
      - couchdb-service

  couchdb-service:
    restart: always
    image: ibmcom/couchdb3
    environment:
      - COUCHDB_PASSWORD=${COUCH_DB_PASSWORD}
      - COUCHDB_USER=${COUCH_DB_USER}
    ports:
      - "${COUCH_DB_PORT}:5984"
    volumes:
      - couchdb3_data:/opt/couchdb/data

  couch-init:
    image: curlimages/curl
    environment:
      PUT_CALL: "curl -u ${COUCH_DB_USER}:${COUCH_DB_PASSWORD} -X PUT couchdb-service:5984"
    depends_on:
      - couchdb-service
    command: ["sh","-c","sleep 10 && $${PUT_CALL}/_users && $${PUT_CALL}/_replicator; fg;"]
  
  redis-service:
    restart: always
    image: redis
    command: redis-server --requirepass ${REDIS_PASSWORD}
    ports:
      - "${REDIS_PORT}:6379"
    volumes:
      - redis_data:/data
  
  watchtower-service:
    image: containrrr/watchtower
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command: --debug --http-api-update bbapps bbworker
    environment:
      - WATCHTOWER_HTTP_API=true
      - WATCHTOWER_HTTP_API_TOKEN=budibase
      - WATCHTOWER_CLEANUP=true
    labels:
      - "com.centurylinklabs.watchtower.enable=false"
    ports:
      - 6161:8080


volumes:
  couchdb3_data:
    driver: local
  minio_data:
    driver: local
  redis_data:
    driver: local
```

**Reply by shogunpurple** [reply]
This looks fine @sergey-dme. It's odd that you would be unauthorized because the dockerhub repo is public. Is it still happening to you now?

**Reply by sergey-dme** [reply]
@shogunpurple Yep, just tried to update bb. there are the same messages in the watchover log

**Reply by KernOil** [reply]
@sergey-dme I was doing the same thing a while back using the command line to update BB when they first moved to the server-based application. it ate the space in my VM like no other. have not had the problem since using the UI updater

**Reply by sergey-dme** [reply]
Hi @shogunpurple Updating bb through the UI fails with the following:
```
time="2021-08-23T19:52:28Z" level=debug msg="Trying to load authentication credentials." container=/bbapps image="sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:28Z" level=debug msg="No credentials for sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879 found" config_file=/config.json
time="2021-08-23T19:52:28Z" level=debug msg="Got image name: sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:28Z" level=debug msg="Checking if pull is needed" container=/bbapps image="sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:28Z" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"
time="2021-08-23T19:52:29Z" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2021-08-23T19:52:29Z" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2021-08-23T19:52:29Z" level=debug msg="Setting scope for auth token" image=sha256 scope="repository:library/sha256:pull"
time="2021-08-23T19:52:29Z" level=debug msg="No credentials found."
time="2021-08-23T19:52:30Z" level=debug msg="Parsing image ref" host=index.docker.io image=sha256 normalized="docker.io/library/sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879" tag=b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879
time="2021-08-23T19:52:30Z" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/library/sha256/manifests/b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:30Z" level=warning msg="Could not do a head request for \"sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879\", falling back to regular pull." container=/bbapps image="sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:30Z" level=warning msg="Reason: registry responded to head request with \"401 Unauthorized\", auth: \"Bearer realm=\\\"https://auth.docker.io/token\\\",service=\\\"registry.docker.io\\\",scope=\\\"repository:library/sha256:pull\\\",error=\\\"insufficient_scope\\\"\"" container=/bbapps image="sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:30Z" level=debug msg="Pulling image" container=/bbapps image="sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879"
time="2021-08-23T19:52:33Z" level=debug msg="Error pulling image sha256:b4d68e77f6804db937aef80f1641a29ee568157945bd10ed746f1beabf77b879, Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"
time="2021-08-23T19:52:33Z" level=info msg="Unable to update container \"/bbapps\": Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied. Proceeding to next."
time="2021-08-23T19:52:33Z" level=debug msg="Trying to load authentication credentials." container=/bbworker image="sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:33Z" level=debug msg="No credentials for sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9 found" config_file=/config.json
time="2021-08-23T19:52:33Z" level=debug msg="Got image name: sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:33Z" level=debug msg="Checking if pull is needed" container=/bbworker image="sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:33Z" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"
time="2021-08-23T19:52:33Z" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2021-08-23T19:52:33Z" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2021-08-23T19:52:33Z" level=debug msg="Setting scope for auth token" image=sha256 scope="repository:library/sha256:pull"
time="2021-08-23T19:52:33Z" level=debug msg="No credentials found."
time="2021-08-23T19:52:33Z" level=debug msg="Parsing image ref" host=index.docker.io image=sha256 normalized="docker.io/library/sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9" tag=6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9
time="2021-08-23T19:52:33Z" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/library/sha256/manifests/6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:34Z" level=warning msg="Could not do a head request for \"sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9\", falling back to regular pull." container=/bbworker image="sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:34Z" level=warning msg="Reason: registry responded to head request with \"401 Unauthorized\", auth: \"Bearer realm=\\\"https://auth.docker.io/token\\\",service=\\\"registry.docker.io\\\",scope=\\\"repository:library/sha256:pull\\\",error=\\\"insufficient_scope\\\"\"" container=/bbworker image="sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:34Z" level=debug msg="Pulling image" container=/bbworker image="sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9"
time="2021-08-23T19:52:36Z" level=debug msg="Error pulling image sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9, Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"
time="2021-08-23T19:52:36Z" level=info msg="Unable to update container \"/bbworker\": Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login': denied: requested access to the resource is denied. Proceeding to next."
time="2021-08-23T19:52:36Z" level=debug msg="Session done: 2 scanned, 0 updated, 2 failed"
```
It was attempt to update to v0.9.112
docker-compose.yaml has been updated recently while updating to v.0.9.111 few days ago using CLI
New server, fresh installation.

**Reply by shogunpurple** [reply]
Thanks @sergey-dme - so I was actually able to recreate this in my own environment, but the update worked. Did the update definitely not work? Sometimes it just takes a while, but when you refresh, then the new version should be shown.

**Reply by sergey-dme** [reply]
@shogunpurple Yep, I can confirm it. I tried to wait few mins, then refresh the page -- it did not work, the version was as before. 
I'll try to do the same next time updating to the v0.9.113

I'm concerned about the errors in watchover log:
```
Error pulling image sha256:6a48c56672eda321a5e98626c93eff3e5cc2995e0bacd04c222aa38db3baccd9, 
Error response from daemon: pull access denied for sha256, repository does not exist 
or may require 'docker login': denied: requested access to the resource is denied

Unable to update container \"/bbworker\": Error response from daemon: pull access denied for sha256, 
repository does not exist or may require 'docker login': denied: requested access to the resource is denied. 
Proceeding to next.

Session done: 2 scanned, 0 updated, 2 failed
```
